sycl/include/sycl/detail/string.hpp

```diff
@@ -0,0 +1,83 @@
+//==----------------- string.hpp - SYCL standard header file ---------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include <cstring>
+#include <string>
+
+#pragma once
+
+namespace sycl {
+inline namespace _V1 {
+namespace detail {
+
+// This class and detail::string_view class are intended to support
+// different ABIs between libsycl and the user program.
+// This class is not inteded to replace std::string for general purpose usage.
+class string {
+  char *str = nullptr;
+
+public:
+  string() {
+    allocate(1);
+    *str = '\0';
+  }
+
+  string(const std::string_view &strn) {
+    allocate(strn.length() + 1);
+    strcpy(str, strn.data());
```

Comment:
I don't think `std::string_view` is guaranteed to be null-terminated. I suspect what's happening is that we have

```
const char *literal = "hello";  // points to -> 'h', 'e', 'l', 'l', 'o', '\0'
std::string_view s{literal}; // N = 5, points to `literal`
```
and then the copy here goes beyond the memory "accessible" by `s` into '\0' that doesn't belong to it. There might be cases when `std::string_view` is created in such a way that there will be no '\0' symbol and the code would fail.

Please correct me if I missed anything that ensure this code is always valid.

PS I'd use `std::string_view::copy`...

Reply:
std::string_view is just a reference to an existing std::string.
So, I am not aware of any C++ string without ending with '\0'.
@jopperm @victor-eds 
Any comments on this?

Reply:
It is true. `std::string_view` is not guaranteed to be null terminated. Using `std::string_view::copy` to fill the string with the characters from the view and setting the last character to `\0` should work.

Reply:
I also don't think we need a separate `allocate` helper. We probably can use dtor/copy/move constructors and  `Swappable` interface (https://en.cppreference.com/w/cpp/algorithm/swap#Specializations) to implement everything else.

Reply:
Thanks for additional feedback.
I used `std::string_view::copy()` instead of `strcpy()`.
`detail::string` is used to return a std::string that is created by libsycl to the user side.
I cannot remove allocate() since we cannot share memory between libsycl and the user side when they use different ABIs. So, `std::swap()` is not an option.

Reply:
What I meant is something like this: https://gcc.godbolt.org/z/a7hWc9ToW
@victor-eds , @jopperm , do you agree that an implementation like that would be more idiomatic C++?

Reply:
Something like that (+ copy constructor and operator) would be more idiomatic (minor: `delete[]` instead of `delete`).

Regarding `constexpr` and `noexcept`, I'd say the default constructor, the move assignment and constructor and the `swap` function should be `noexcept`. The destructor is `noexcept` by default already, so we're good there.

I don't see much point in making default and move constructors `constexpr` as this class's functionality is limited by design and I don't think we'll be using that in contexts requiring `constexpr`. However, adding `constexpr` does not imply ABI breakage anyway, right?

Reply:
Hi @aelovikov-intel @victor-eds @jopperm 
I think addressed all your feedback.
If I missed anything, please let me know.